### PR TITLE
build: modernize tooling, CI, and pre-commit infrastructure

### PR DESCRIPTION
**Pull request by SudipSinha**
## Summary

Modernize the project's tooling infrastructure: replace legacy linters, update dependencies, reorganize configuration, modernize CI workflows, and update pre-commit hooks.

**Jira:** RHOAIENG-55591, RHOAIENG-55590, RHOAIENG-55592, RHOAIENG-55596

### Design decision: tool installation without CI enforcement

This PR installs and configures the new tools (`ruff`, `pyrefly`, `bandit`) but does **not** add `lint` or `type-check` CI jobs. Main's code doesn't yet comply with these tools — enforcement is deferred to a follow-up PR after source code cleanup PRs land.

### pyproject.toml

**Dependencies:**
- Replace `isort`/`flake8`/`mypy` with `ruff`/`pyrefly`/`bandit[toml]` in dev group
- Add `types` group (`pandas-stubs`, `scipy-stubs`, `scikit-learn-stubs`, `h5py-stubs`)
- Add `pytest-xdist[psutil]>=3.8.0` for future parallel test execution

**New tool configs:** `[tool.ruff]`, `[tool.pytest.ini_options]`, `[tool.pyrefly]` (with broad sub-configs for pre-existing type issues), `[tool.bandit]`

**Removed:** `[tool.isort]`, `[tool.mypy]`, legacy `[tool.ruff.format]`

**Note:** `select = ["ALL"]` with per-file-ignores deferred to strict-linting PR.

### Lock file

Regenerated `uv.lock` for updated dependency groups.

### Test fixes

- Fix pandas compat: remove deprecated `axis=1` from `df.drop(columns=...)` in `test_fairness.py`
- Modernize RNG: replace `np.random.seed`/`np.random.*` with `np.random.default_rng`
- Add type annotations, docstrings, hypothesis `assume()` guards

### Pre-commit hooks

**Removed:** `flake8`, `mypy`, `isort`

**Added:** `ruff-check`, `ruff-format`, `pyrefly` (local hook), `bandit`, `check-yaml`, `conventional-pre-commit`

**Updated:** ruff v0.15.10, gitleaks v8.30.1

### CI workflows

**python-tests.yaml:**
- Modernize: `uv` replaces `pip`+`venv`, `setup-uv@v4` with caching
- Add Python 3.14 to matrix with `allow-prereleases: true`
- Add concurrency group with `cancel-in-progress`
- Remove dead `cache: pip`
- No lint/type-check jobs (deferred), no `-n auto` (MariaDB tests need xdist_group markers first)

**security-scan.yaml:**
- Switch to `uv` for dependency installation
- Add concurrency group

**build-and-push.yaml → split into three workflows (security rework):**
- `build-and-push.yaml`: simplified to `push`-only (main/tags), all PR logic removed
- `ci-build.yaml` (new): `pull_request` trigger, builds image with NO secret access, saves as artifact
- `ci-publish.yaml` (new): `workflow_run` trigger, loads artifact, pushes to Quay, generates manifests, posts PR comment
- Security: all `${{ }}` in `run:` steps moved to `env:` blocks; `docker login` uses `--password-stdin`
- UX: Commit Status API reports CI Publish result back to the PR checks
- Fix `Dockerfile` → `Containerfile` reference (pre-existing bug)

### Design decision: two-workflow split for PR image builds

The original `build-and-push.yaml` used `pull_request_target` with a patch-apply mechanism to build PR images. This was fragile (`git apply -3` failures, `fetch-depth: 0`, patch validation) and still allowed untrusted Containerfile `RUN` commands to execute with secret access.

The rework splits into `ci-build.yaml` (no secrets) + `ci-publish.yaml` (secrets, no untrusted code). This is GitHub's recommended pattern for `pull_request_target` security.

**Tradeoffs:**

| Aspect | Before (single workflow) | After (two-workflow split) |
|---|---|---|
| Security | Untrusted code runs with secret access (mitigated by label gate + base SHA checkout) | Untrusted code never has secret access (architectural guarantee) |
| PR Checks tab | Single check visible | CI Build visible; CI Publish reports via Commit Status API |
| Artifact transfer | N/A (single job) | Image tarball uploaded/downloaded between workflows |
| Trivy SARIF | Associated with PR ref | Associated with default branch ref (limitation of `workflow_run`) |
| Branch protection | `build-and-push-ci` check name | New check names: `CI Build / build` and `CI Publish` — **update required status checks after merge** |
| Bootstrap | N/A | `ci-publish.yaml` must exist on main before `workflow_run` triggers — first CI Build after merge is the first to trigger CI Publish |

### Deferred to follow-up PRs

| Item | Why | Follow-up |
|---|---|---|
| `lint`/`type-check` CI jobs | Main's code doesn't pass ruff/pyrefly | After source code cleanup PRs |
| `select = ["ALL"]` + per-file-ignores | Same — code must comply first | After source code cleanup PRs |
| `-n auto` parallel tests | MariaDB tests share DB, cause race conditions | After adding `@pytest.mark.xdist_group` markers |
| Tighten broad pyrefly sub-configs | Temporary workaround for pre-existing type issues | After source code cleanup PRs |

### Known CI status

| Check | Status | Notes |
|---|---|---|
| test (3.12) | ✅ | |
| test (3.14) | ✅ | |
| bandit-scan | ✅ | |
| CodeRabbit | ✅ | |
| CI Build | ✅ new | Replaces `build-and-push-ci`. Uses `pull_request` trigger. |
| CI Publish | N/A | `workflow_run` trigger — only activates after this PR merges to main (bootstrap). |
| build-and-push-ci | ❌ expected | Still runs main's workflow (which has the old `Dockerfile` ref). Replaced by CI Build + CI Publish on merge. |
| pre-commit.ci | ❌ expected | Pre-existing failure on main. |
| GHAS/Bandit | ⚠️ warning | "Configuration not found" — workflow changed, GHAS can't diff against main's baseline. Resolves on merge. |

## Test plan

- [x] `uv sync --all-groups` succeeds
- [x] `uv run pytest tests/ -v` — all tests pass
- [x] `uv run pyrefly check src/` — 0 errors
- [x] `uv run ruff check tests/core/metrics/test_fairness.py` — 0 violations
- [x] `pre-commit run --all-files` — all hooks pass locally
- [x] CI: test (3.12) + test (3.14) + bandit-scan green

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Modernize local tooling, test utilities, and CI workflows around uv while updating security-related dependencies and pre-commit hooks.

New Features:
- Introduce pyrefly type checking configuration with project- and path-specific suppressions.
- Add bandit configuration and integration into pre-commit and security scanning workflows.

Bug Fixes:
- Fix database build-and-push workflow to reference the correct Containerfile instead of a non-existent Dockerfile.

Enhancements:
- Refine fairness metric tests with modern NumPy RNG usage, explicit type hints, and stronger property-based assumptions to reduce flakiness and improve clarity.
- Exclude generated protobuf stubs and proto paths from static analysis to avoid noisy findings.
- Tune pytest configuration for strict asyncio handling and future-friendly grouping of parallel tests.

Build:
- Update Python dependency groups, including dev, test, and notebook extras, with newer versions and added stub packages, then regenerate the uv lockfile.

CI:
- Migrate test and security workflows to use setup-uv with cached dependency resolution, add concurrency controls, expand the Python test matrix to include 3.14, and fix container builds to use the Containerfile with updated checkout actions.

Deployment:
- Align container build workflow with the Containerfile and ensure CI-tagged images are built and labeled consistently.

Tests:
- Improve fairness metric test utilities and providers, adding annotations and docstrings, and adjust strategies and guards to ensure stable, meaningful property-based checks.

## Summary by CodeRabbit

* **Chores**
  * Modernized developer tooling and dependency groups, tightened runtime/test/notebook minimums, added typed-stub group, and simplified lint/type/security configs.
  * Reworked CI: added dedicated CI build/publish workflows, simplified push-based image build/push, standardized concurrency/cancellation, and integrated image scanning/reporting.
  * Updated pre-commit hooks and added commit-message validation.

* **Tests**
  * Strengthened tests with added typing, stabilized randomness, improved property-based guards, and tightened assertions.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -18,26 +18,38 @@ dependencies = [
...
 ]
 
 [dependency-groups]
-dev = ["isort>=7,<8", "flake8>=7,<8", "mypy>=1.5,<2", "httpx>=0.25,<0.29"]
+dev = [
+    "ruff>=0.15.10",
+    "pyrefly>=0.61",
+    "bandit[toml]>=1.9.4",
+    "httpx>=0.28.1",
+]
```
[the remark this answers is not in the document]
Comment (SudipSinha) [reply pyproject.toml:33]: False positive — ruff 0.15.10 is installed and working. CI passes with it. The version check was run against stale PyPI data (ruff releases frequently; 0.15.10 was released well before this PR).

```
$ uv pip show ruff | grep Version
Version: 0.15.10
```

File: pyproject.toml [hunk 2]
```diff
@@ -18,26 +18,38 @@ dependencies = [
...
-    "pytest-cov>=7,<8",
-    "pytest-asyncio>=1,<2",
-    "hypothesis>=6,<7",
-    "aif360>=0.6,<0.7",
-    "pillow>=12.1",          # Security: Resolves CVE-2026-25990 (HIGH - Out-of-bounds write)
+    "pytest>=9.0.3",
+    "pytest-cov>=7.1.0",
+    "pytest-asyncio>=1.3.0",
+    "pytest-xdist[psutil]>=3.8.0",
+    "hypothesis>=6.151.12",
```
[the remark this answers is not in the document]
Comment (SudipSinha) [reply pyproject.toml:45]: False positive — hypothesis 6.151.14 is installed and working (constraint is >=6.151.12). Hypothesis uses calver and releases very frequently. The version check was run against stale PyPI data.

```
$ uv pip show hypothesis | grep Version
Version: 6.151.14
```

File: pyproject.toml [hunk 3]
```diff
@@ -59,34 +71,267 @@ include = ["src"]
...
+# Test utilities use many parameters
+"tests/service/data/test_utils.py" = ["PLR0913"]
 
 [tool.pytest.ini_options]
-testpaths = ["tests"]
-python_files = "test_*.py"
-python_functions = "test_*"
+asyncio_mode = "strict"
 asyncio_default_fixture_loop_scope = "function"
+addopts = "--dist=loadgroup"
```
[the remark this answers is not in the document]
Comment (SudipSinha) [reply pyproject.toml:84]: Valid observation, but this is pre-existing code — we didn't modify those test files in this PR.

This is already tracked as a deferred item. We removed `-n auto` from CI specifically because of this issue (MariaDB tests race when parallelized). The follow-up PR will add `@pytest.mark.xdist_group("mariadb")` markers and proper `tearDown` cleanup before re-enabling parallel execution.

**Comment by SudipSinha** [comment]
## Addressing Sourcery review feedback

**1. pyrefly `python-version = "3.14.0"` vs mypy's 3.12**

Intentional. The config already has a comment: `# Forward-looking: pyrefly uses this for type checking, not runtime`. The project supports Python 3.12–3.14 (`requires-python = ">=3.12, <3.15"`). Setting pyrefly to the highest supported version ensures type checking catches forward-compatibility issues. This is different from mypy's `python_version` which constrained to the minimum.

**2. Dependencies lost upper bounds (e.g. pytest, httpx)**

These are dev/test dependencies, not runtime. The `uv.lock` file pins exact versions for reproducibility. Upper bounds on dev tools are overly restrictive — they prevent picking up bug fixes and security patches. The runtime dependencies retain their upper bounds (e.g. `scikit-learn>=1.7,<2`, `fastapi>=0.116,<0.129`).

**3. `utilities/manifests` no longer excluded from ruff**

The `utilities/manifests` directory does not exist in this repo — the old `extend-exclude` was dead config. Verified:
```
$ ls -d utilities/manifests
ls: utilities/manifests: No such file or directory
```
The new `exclude` targets the only generated file that actually needs exclusion: `src/proto/grpc_predict_v2_pb2.py`.

File: .github/workflows/build-and-push.yaml [hunk 4]
```diff
@@ -44,11 +50,11 @@ jobs:
           mode: minimum
           count: 1
           labels: "ok-to-test, lgtm, approved"
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
         if: env.BUILD_CONTEXT == 'ci'
         with:
           ref: ${{ github.event.pull_request.head.sha }}
```
[the remark this answers is not in the document]
Comment (RobGeada) [reply .github/workflows/build-and-push.yaml]: Yep - this is an explicit directive from RH op-sec too.

**Review by RobGeada — CHANGES_REQUESTED** [review]

File: .github/workflows/build-and-push.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (SudipSinha) [reply .github/workflows/build-and-push.yaml]: **Security fix applied.** The `pull_request_target` workflow now checks out `github.event.pull_request.base.sha` instead of the untrusted PR head SHA. PR code changes are fetched via the GitHub API and applied as a git patch with `git apply -3`. Only trusted base branch code has access to secrets (QUAY_ROBOT_SECRET, TRUSTYAI_CI_BOT_SSH_KEY).

File: pyproject.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (SudipSinha) [reply pyproject.toml:84]: Yes

File: .github/workflows/build-and-push.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (SudipSinha) [reply .github/workflows/build-and-push.yaml]: **Addressed all three points:**

1. **`fetch-depth: 0`** added to actions/checkout@v4 so `git apply -3` has full history to find the merge base.
2. **`curl -sSf --fail-with-body`** replaces `curl -s` — fails on HTTP errors with `-f`/`--fail-with-body`, and shows error details with `-S`.
3. **Patch validation** added: greps for `^diff --git` header before `git apply -3`, prints the file content on failure for debugging.

File: .github/workflows/python-tests.yaml [hunk 5]
```diff
@@ -21,48 +26,42 @@ jobs:
...
+        with:
+          enable-cache: true
+          cache-dependency-glob: "uv.lock"
 
-      - name: Install protobuf compiler
+      - name: Install system dependencies
         run: |
           sudo apt-get update
-          sudo apt-get install -y protobuf-compiler
+          sudo apt-get install -y libmariadb3 libmariadb-dev mariadb-server protobuf-compiler
```
[the remark this answers is not in the document]
Comment (SudipSinha) [reply .github/workflows/python-tests.yaml]: **Fixed.** Replaced `mariadb-server` with `mariadb-client` to avoid port 3306 conflict. Added `sudo service mariadb stop || true` alongside the existing `mysql` stop command to shut down both services before the MariaDB action runs. The `|| true` makes both tolerant of missing services.

File: .github/workflows/security-scan.yaml [hunk 6]
```diff
@@ -23,30 +28,24 @@ jobs:
...
-          python -m pip install bandit[toml]
+        run: uv sync --group dev
 
       - name: Run Bandit Security Scan
         uses: PyCQA/bandit-action@v1
         with:
           targets: "src/"
+          # Note: configFile omitted — bandit-action bundles Python 3.8 without
+          # TOML parser. Skips from [tool.bandit] in pyproject.toml are applied
+          # when running bandit locally/in pre-commit (which uses our venv).
```
[the remark this answers is not in the document]
Comment (SudipSinha) [reply .github/workflows/security-scan.yaml]: **Fixed.** Replaced `PyCQA/bandit-action@v1` with `uv run bandit -r src -c pyproject.toml -f sarif -o results.sarif` so the `[tool.bandit]` config (skips, exclusions) is applied consistently with pre-commit. The action bundles Python 3.8 without a TOML parser and ignores `pyproject.toml` by default.

File: tests/core/metrics/test_fairness.py [hunk 7]
```diff
@@ -10,114 +14,159 @@
...
-def truth_predict_output(X: pd.DataFrame = X, y: pd.Series = y):
+def truth_predict_output(
+    X: pd.DataFrame = X,
+    y: pd.Series = y,
+) -> tuple[pd.Series, pd.DataFrame]:
+    """Generate ground truth and predictions indexed by gender for fairness testing."""
     y.index = X["Gender"]
     y_pred = pd.DataFrame(train_model())
     y_pred.index = X["Gender"]
     return y, y_pred
```
[the remark this answers is not in the document]
Comment (SudipSinha) [reply tests/core/metrics/test_fairness.py]: **Fixed.** `train_model()` now operates on `X_local = X.copy()` instead of mutating the shared fixture. `truth_predict_output()` captures `gender` and `y_true` as copies before calling `train_model()`, so the original `X`/`y` fixtures remain unmodified and protected-attribute labels stay aligned between `y_true` and `y_pred`.

File: .github/workflows/build-and-push.yaml [hunk 8]
```diff
@@ -19,6 +19,12 @@ on:
       - '**.adoc'
       - '*.txt'
     types: [labeled, opened, synchronize, reopened]
+
+# Cancel in-progress PR builds; never cancel main (could corrupt image push)
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: ${{ github.event_name == 'pull_request' }}
+
```
[the remark this answers is not in the document]
Comment (SudipSinha) [reply .github/workflows/build-and-push.yaml]: Fixed. Changed `github.event_name == 'pull_request'` to `github.event_name == 'pull_request_target'`. For `pull_request_target` triggers, `github.event_name` evaluates to the trigger type, so the original condition was always false.

**Comment by SudipSinha** [comment]
@coderabbitai review

**Comment by SudipSinha** [comment]
@coderabbitai review

Two additional minor fixes applied per your latest review:
- Added `DATABASE_ATTEMPT_MIGRATION` to environment snapshot in Maria tests
- Made tearDown more robust by ensuring environment restoration happens even if cleanup fails

**Comment by SudipSinha** [comment]
Addressed the final CodeRabbit review thread (deterministic favorable-rate helper) in commit `a1ba0f9`:
- Added RNG seed parameter to `set_favorable_outcomes()` for reproducible test data
- Relaxed tolerance in DIR equal-rates test to account for discretization effects

The environment approval gate issue (#2) has been tracked separately as #129 since it's a security enhancement requiring policy decision (not a bug fix for this PR).

**Comment by SudipSinha** [comment]
## Response to CodeRabbit review threads

### 1. Edge case tests for DIR (test_fairness.py, line null)
**Status:** Already tracked in #128  
**Rationale:** This is test expansion (adding new test cases for zero-favorable scenarios), not a bug fix. Deferred to follow-up work as documented in the issue.

---

### 2. Environment approval gate (ci-publish.yaml:17) 
**Status:** Tracked in #129  
**Rationale:** This is a security enhancement requiring organizational policy decision about required reviewers for the `CI Publish` workflow. Created dedicated issue with security context and implementation guidance.

---

### 3. Deterministic favorable-rate helper (test_fairness.py:233)
**Status:** ✅ Fixed in a1ba0f9  
**Changes:**
- Added `seed` parameter to `set_favorable_outcomes()` for deterministic RNG
- Relaxed DIR equal-rates tolerance from `1e-5` to `0.01` to account for discretization effects from `round(len(group) * target_rate)`

All tests pass with deterministic behavior.

---

All actionable items for this PR are now addressed. The two deferred items (#128, #129) are tracked separately as they represent new work beyond the scope of this tooling modernization PR.

**Review by RobGeada — APPROVED** [review]
